### PR TITLE
User defined variables

### PR DESCRIPTION
**Pull request by okonuskan**
Added possibility for users to inject their own custom variables into
the redux-responsive state. The users has access to all available data and can customise the reducer to fully meet the needs of their own
projects.

### CODE REVIEW

**Comment by AlecAivazis** [comment]
@okonuskan thanks for submitting this PR! I especially appreciate that you added a section to the Readme. Unfortunately, I'm having a hard time understanding the use-case this feature supports - can you give me an example?

**Comment by okonuskan** [comment]
I tried to explain the reasoning behind this feature in the provided readme section. I wanted to do something like `browser.is.small` instead of `browser.mediaType === 'small'` basically because it looks better :) The reason I built this as a customizeable feature instead of hard coding the desired variables into the state reducer was that I saw more use cases where a user like me might want full access to the inner workings of this repo. Here are some examples:

- Enables testing with `browser.is.small` instead of `browser.mediaType === 'small'`

- The responsiveness code in most of the projects I have worked on almost exclusively distinguish between mobile and desktop. Instead of keeping track of all the breakpoint sizes, my PR would enable the project to use only two custom variables called `browser.isMobile` and `browser.isDesktop`, or even one of them assuming desktop when `browser.isMobile === false`.

- Using `lessThan` or `greaterThan` is good but what if you need `lessThanOrEqualTo` / `greaterThanOrEqualTo` (I would call them `max` / `min`)?

I am of course aware that the same information can be derived in the redux-connect statement or in the component itself, but I prefer keeping this kind of repetitive code in one place - outside of my component files.

To sum it up: This feature lets the user define custom viewport variables that suits their needs and coding style.

I thought about creating a PR with new variables in the reducer, but seeing that it already was pretty crowded I thought this approach might fit the project better, besides being customizeable.

**Comment by AlecAivazis** [comment]
OO i really like `browser.is` - that's really clean. I might end up renaming `mediaType` to that in the next version. 

Thanks for writing up that response - I see how having computed values in the browser state would be very powerful. I'm definitely interested in merging this PR and adding the feature - thanks for keeping things so extendible!

While I have your attention, can you think of nice way to add this kind of feature for browser events? I'm interested in removing `height` and `width` since they are very generic and only a small subset of the attributes/events someone might be interested. This means I need a good way to add listeners for browser events and a way to inject things into the state (which you've done here) - curious if you have any ideas on how to handle the event handler side of this problem since you've gotten the harder half

**Comment by okonuskan** [comment]
Do I understand you correctly if you want to enable users to extend the
library in order to hook into other browser events?

I might be wrong, but I think all relevant event handling should take place
within this library and all viewport data to the reducer and the
"customVariableConfig function".

I am on the move at the moment and will look into browser.is later.
Meanwhile maybe you could clarify what you wanted help with?

/Olcay

**Comment by AlecAivazis** [comment]
So maybe more context would help: When I added `performanceMode`, I wanted to have it be the standard behavior however, it affects the way `height` and `width` can be treated since they would only update when the current media type changed. I thought about removing them (since they can be grabbed from the window object) but there are people who are relying on having a browser resize.

So one way to fix that (which you've basically just added support for), is to remove the `width` and `height` from the default reducer, and then provide a way for the users to have the width and height added to the state. Then, all the user has to do is attach an event handler that recalculates the state on resize, and they have the old behavior. I suspect the number of people who need the fine grain height/width is much smaller than the people who would use `performanceMode`

**Comment by AlecAivazis** [comment]
`browser.is` and the change of `height`/`width` would have to come in new versions so we should get your PR in before we start changing the API. I'll review this over lunch today and hopefully we can publish it very soon

**Comment by okonuskan** [comment]
Looking at addEventHandlers.js, basically what you doing is exposing the store to the event handlers addPerformanceModeHandlers() or addThrottledHandlers() depending on configuration of performance mode. I would suggest something like passing an array of multiple [someName]Handlers() as the only argument to createResponsiveStoreEnhancer(). addPerformanceModeHandlers would be my first choice too as the default handler, but addThrottledHandlers could still be kept in the project and referenced as a separate handler that can be used together with createResponsiveStoreEnhancer().

However I think this ocercomplicated the setup a little. The main two problems I see are:

1. The performanceMode configuration is a simple and elegant solution. I really tried to see a use case with other event handlers than the two included, but I couldn't think of any other handler. Also there isn't anything preventing an advanced user from skipping the storeEnhancer and implementing a event handler which runs dispatch(calculateResponsiveState(window)) outside of redux-responsive (assuming calculateResponsiveState() can be imported).

2. Redux users are used to writing reducers so I think my PR would be easy to understand, but injecting event handlers into the createResponsiveStoreEnhancer() won't be in many persons comfort zones.

I would either remove the throttle-handler completely and show users how they can accomplish the same thing setting up their own throttled event handler which fires dispatch(calculateResponsiveState(window)), or I would just keep it as it is today.

**Comment by AlecAivazis** [comment]
Wow @okonuskan thanks for a very complete response. I think you are right - I'm in favor of removing the `height` and `width` now that you've added the ability for users to inject custom data types to the state. It also simplifies the code base susbstantially so that's a definite win 👍

**Comment by okonuskan** [comment]
Ok! Glad I could be of any help :)

I actually had a simililar script like yours running in my projects, but I
never perfected it like you did. The reason I stumbled upon this project
was that I had to rewrite my own code to stop it from updating the props
too often. Really great work with the performance optimization! Appreciate
it.

If you want to get feedback on future ideas, don't hesitate to contact me.

/Olcay

**Comment by AlecAivazis** [comment]
I will definitely keep that in mind - thanks! 

So while reviewing your PR, it became clear that since we need to pass an additional argument to the reducer factory, it would be better to just replace the `infinity` argument with an object, similar to how the store enhancer factory works. Since that's an API breaking change, I decided to just use your PR as the basis for the new version. That work is happening [here](https://github.com/AlecAivazis/redux-responsive/tree/4.0.0). I still have to remove `height` and `width` as well as implement `is` but I am working on that now and should have it done very soon. I'll keep you up to date and let you know when that's ready for review :+1:

**Comment by AlecAivazis** [comment]
@okonuskan i finished adding all of the changes I had in mind for the new version. I'll submit a PR now for you to explore but I want to update the README with the changes I've made so I'm going to hold off on merging until I do so.
